src/services/implementation/ClientServiceImpl.java

```diff
@@ -0,0 +1,70 @@
+package services.implementation;
+
+import entities.Client;
+import enums.StatusEnum;
+import exceptions.services.EntityNotFoundException;
+import exceptions.services.IdNullPointerException;
+import exceptions.services.NullEntityException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import services.interfaces.IClientInterface;
+import util.ClientServiceUtility;
+
+public class ClientServiceImpl implements IClientInterface {
+  private List<Client> clients;
+  private Long currentId;
+
+  public ClientServiceImpl() {
+    this.clients = new ArrayList<>();
+    this.currentId = 0L;
+  }
+
+  @Override
+  public Optional<Client> insert(Client client) {
+    client.setId(this.currentId);
+    client.setCode(ClientServiceUtility.createCode(this.currentId));
+    this.currentId++;
+    this.clients.add(client);
+    return Optional.of(client);
+  }
+
+  @Override
+  public Optional<Client> findById(Long id) {
+    Optional.ofNullable(id).orElseThrow(IdNullPointerException::new);
+    Client client =
+        this.clients.stream()
+            .filter(c -> id.equals(c.getId()))
+            .findAny()
+            .orElseThrow(
+                () -> new EntityNotFoundException(Client.class.getClass().getSimpleName(), id));
+    return Optional.ofNullable(client);
+  }
+
+  @Override
+  public Optional<Client> delete(Long id) {
+    Optional.ofNullable(id).orElseThrow(IdNullPointerException::new);
+    Client client = this.findById(id).get();
+    client.setStatus(StatusEnum.DELETED);
+    this.update(client);
+    return Optional.of(client);
```

Comment:
`this.update` already returns an optional.

Reply:
solved